doc/index.rst

```diff
@@ -0,0 +1,599 @@
+
+.. data:: ClassVar
+
+   See :py:data:`typing.ClassVar`. In ``typing`` since 3.5.3.
+
+.. data:: Concatenate
+
+   See :py:data:`typing.Concatenate`. In ``typing`` since 3.10.
+
+.. data:: Final
+
+   See :py:data:`typing.Final`. In ``typing`` since 3.8.
+
+.. data:: Literal
+
+   See :py:data:`typing.Literal`. In ``typing`` since 3.8.
+
+   :py:data:`typing.Literal` does not flatten or deduplicate parameters on Python <3.9.1, and a
+   caching bug was fixed in 3.10.1/3.9.8. The ``typing_extensions`` version
+   flattens and deduplicates parameters on all Python versions, and the caching
+   bug is also fixed on all versions.
+
+   .. versionchanged:: 4.6.0
+
+      Backported the bug fixes from :py:gh:`29334`, :py:gh:`23294`, and :py:gh:`23383`.
+
+.. data:: LiteralString
+
+   See :py:data:`typing.LiteralString`. In ``typing`` since 3.11.
+
+   .. versionadded:: 4.1.0
+
+.. class:: NamedTuple
+
+   See :py:class:`typing.NamedTuple`.
+
+   ``typing_extensions`` backports several changes
+   to ``NamedTuple`` on Python 3.11 and lower: in 3.11,
+   support for generic ``NamedTuple`` was added, and
+   in 3.12, the ``__orig_bases__`` attribute was added.
+
+   .. versionadded:: 4.3.0
+
+      Added to provide support for generic ``NamedTuple``.
+
+   .. versionchanged:: 4.6.0
+
+      Support for the ``__orig_bases__`` attribute was added.
+
+.. data:: Never
+
+   See :py:data:`typing.Never`. In ``typing`` since 3.11.
+
+   .. versionadded:: 4.1.0
+
+.. class:: NewType(name, tp)
+
+   See :py:class:`typing.NewType`. In ``typing`` since 3.5.2.
+
+   ``NewType`` was made picklable in 3.10 and an error message was
+   improved in 3.11; ``typing_extensions`` backports these changes.
+
+   .. versionchanged:: 4.6.0
+
+      The improvements from Python 3.10 and 3.11 were backported.
+
+.. data:: NoReturn
+
+   See :py:data:`typing.NoReturn`. In ``typing`` since 3.5.4 and 3.6.2.
+
+.. data:: NotRequired
+
+   See :py:data:`typing.NotRequired`. In ``typing`` since 3.11.
+
+   .. versionadded:: 4.0.0
+
+.. class:: ParamSpec(name, *, default=...)
+
+   See :py:class:`typing.ParamSpec`. In ``typing`` since 3.10.
+
+   The ``typing_extensions`` version adds support for the
+   ``default=`` argument from :pep:`696`.
+
+   .. versionchanged:: 4.4.0
+
+      Added support for the ``default=`` argument.
+
+   .. versionchanged:: 4.6.0
+
+      The implementation was changed for compatibility with Python 3.12.
+
+.. class:: ParamSpecArgs
+
+.. class:: ParamSpecKwargs
+
+   See :py:class:`typing.ParamSpecArgs` and :py:class:`typing.ParamSpecKwargs`.
+   In ``typing`` since 3.10.
+
+.. class:: Protocol
+
+   See :py:class:`typing.Protocol`. In ``typing`` since 3.8.
+
+   Python 3.12 improves the performance of runtime-checkable protocols;
+   ``typing_extensions`` backports this improvement.
+
+   .. versionchanged:: 4.6.0
+
+      Backported the ability to define ``__init__`` methods on Protocol classes.
+
+   .. versionchanged:: 4.6.0
```

Comment:
Could possibly also mention https://github.com/python/typing_extensions/pull/140 and #132 in versionchanged notes for `Protocol` — they bring the implementation in line with Python 3.12, and they're strictly-speaking bugfixes, but they're also minor compatibility breaks with previous versions

Reply:
Done, thanks for all the comments!